**Parcial 1
Juana Valeria Pinzon. 202112577
Alan Romero. 202212891**


**a.** Implemente el algoritmo de Muller para encontrar la raíz de la función:
f(x) = e^−x − x 

In [40]:
import numpy as np 
tol=0.00001

def f(x):
    expresion = -x+np.e**(-x)
    return expresion

def muller(x0, x1, x2, tol):
    h1 = x1 - x0
    h2 = x2 - x1
    d1 = (f(x1) - f(x0)) / h1
    d2 = (f(x2) - f(x1)) / h2
    d = (d2 - d1) / (h2 + h1)
    i = 0
    while abs(h2) > tol:
        b = d2 + h2 * d
        D = np.sqrt(b * b - 4 * f(x2) * d)
        if abs(b - D) < abs(b + D):
            E = b + D
        else:
            E = b - D
        h = -2 * f(x2) / E
        p = x2 + h
        x0 = x1
        x1 = x2
        x2 = p
        h1 = x1 - x0
        h2 = x2 - x1
        d1 = (f(x1) - f(x0)) / h1
        d2 = (f(x2) - f(x1)) / h2
        d = (d2 - d1) / (h2 + h1)
        i += 1
    return p, i

root, iterations = muller(0, 0.5, 1, tol)
print(f"La raíz encontrada es: {root:.6f}")
print(f"Se necesitaron {iterations} iteraciones para encontrar la raíz")

La raíz encontrada es: 0.567143
Se necesitaron 3 iteraciones para encontrar la raíz


**Respuesta:**
 La raíz encontrada es: 0.567143
Se necesitaron 3 iteraciones para encontrar la raíz.

**b.** Use el teorema de Boltzano para determinar dos valores x0, x1 tal que f(x0)f(x1) < 0

In [41]:
def has_sign_change(f, a, b):
    return f(a) * f(b) < 0

# Función para encontrar los valores x0 y x1
def find_x0_x1(f, a, b, tol):
    if not has_sign_change(f, a, b):
        return None, None

    while (b - a) / 2.0 > tol:
        c = (a + b) / 2.0
        if f(c) == 0:
            return c, c
        elif f(c) * f(a) < 0:
            b = c
        else:
            a = c

    return a, b

x0, x1 = find_x0_x1(f, 0, 1, tol)

if x0 is not None and x1 is not None:
    print("Los valores x0 y x1 son:", x0, x1)
else:
    print("No se encontró cambio de signo en el intervalo [a, b].")

Los valores x0 y x1 son: 0.567138671875 0.5671539306640625


**Respuesta:**
Los valores x0 y x1 son: 0.567138671875 0.5671539306640625


**c.** Defina x2 como el punto medio de x0 y x1.

In [42]:
# Calcula x0 y x1 usando la función find_x0_x1
x0, x1 = find_x0_x1(f, 0, 1, tol)

# Calcula x2 como el punto medio entre x0 y x1
x2 = (x0 + x1) / 2

if x0 is not None and x1 is not None:
    print("Los valores x0 y x1 son:", x0, x1)
    print("El valor x2 como punto medio entre x0 y x1 es:", x2)
else:
    print("No se encontró cambio de signo en el intervalo [a, b].")


Los valores x0 y x1 son: 0.567138671875 0.5671539306640625
El valor x2 como punto medio entre x0 y x1 es: 0.5671463012695312


**Respuesta:**
Los valores x0 y x1 son: 0.567138671875 0.5671539306640625
El valor x2 como punto medio entre x0 y x1 es: 0.5671463012695312


**d.** Encuentre x3 según el algoritmo.

In [47]:
def muller2(x0, x1, x2):
    h1 = x1 - x0
    h2 = x2 - x1
    d1 = (f(x1) - f(x0)) / h1
    d2 = (f(x2) - f(x1)) / h2
    d = (d2 - d1) / (h2 + h1)
    b = d2 + h2 * d
    D = np.sqrt(b * b - 4 * f(x2) * d)
    if abs(b - D) < abs(b + D):
        E = b + D
    else:
        E = b - D
    h = -2 * f(x2) / E
    x3 = x2 + h
    return x3

x3 = muller2(x0, x1, x2)
print(f"La nueva aproximación es: {x3:.6f}")


La nueva aproximación es: 0.567143


**Respuesta:**
El valor de x3 es: 0.567143

**e.** Use el siguiente criterio de parada:
$\epsilon = |f(x3)|< 1 × 10^{-10}$
La rutina se debe detener si no converge en 100 iteraciones

In [ ]:
def nuevo_muller(x0, x1, x2, tol, max_iter):
    h1 = x1 - x0
    h2 = x2 - x1
    d1 = (f(x1) - f(x0)) / h1
    d2 = (f(x2) - f(x1)) / h2
    d = (d2 - d1) / (h2 + h1)
    i = 0
    while abs(f(x2)) > tol and i < max_iter:
        b = d2 + h2 * d
        D = np.sqrt(b * b - 4 * f(x2) * d)
        if abs(b - D) < abs(b + D):
            E = b + D
        else:
            E = b - D
        h = -2 * f(x2) / E
        p = x2 + h
        x0 = x1
        x1 = x2
        x2 = p
        h1 = x1 - x0
        h2 = x2 - x1
        d1 = (f(x1) - f(x0)) / h1
        d2 = (f(x2) - f(x1)) / h2
        d = (d2 - d1) / (h2 + h1)
        i += 1
    if i == max_iter:
        print("El método no convergió en el número máximo de iteraciones")
    return p, i

root, iterations = nuevo_muller(x0, x1, x2, tol, 100)
print(f"La raíz encontrada es: {root:.6f}")
print(f"Se necesitaron {iterations} iteraciones para encontrar la raíz")

**Respuesta:**
 La raiz encontrada es: 0.567143